fpga_interchange/arch_pack_chains.cc

```diff
@@ -0,0 +1,237 @@
+    }
+    return ci;
+}
+
+bool getClusterPlacement(ClusterId cluster, BelId root_bel,
+                         std::vector<std::pair<CellInfo *, BelId>> &placement) const
+{
+    // Load coord config if available?
+    // Place cluster
+    NPNR_ASSERT_FALSE("unimplemented");
+}
+
+// TODO
+//ArcBounds getClusterBounds(ClusterId cluster) const
+//{
+//}
+//Loc getClusterOffset(const CellInfo *cell) const
+//{
+//}
+//bool isClusterStrict(const CellInfo *cell) const
+//{
+//}
+
+void dump_chains(const ChipInfoPOD *chip_info, Context *ctx)
+{
+    std::unordered_map<IdString, const BelChainPOD *> bel_chain_prototypes;
+    for (size_t i = 0; i < chip_info->bel_chains.size(); ++i) {
+        const auto &bel_chain = chip_info->bel_chains[i];
+        IdString bel_chain_name(bel_chain.name);
+        bel_chain_prototypes.emplace(bel_chain_name, &bel_chain);
+        log_info("BEL chain '%s' loaded! Parameters:\n", bel_chain_name.c_str(ctx));
+        log_info("  - sites:\n");
+            for (auto site : bel_chain.sites) {
+                IdString site_name(site);
+                log_info("      - %s\n", site_name.c_str(ctx));
+            }
+        log_info("  - cells:\n");
+            for (auto cell : bel_chain.cells) {
+                IdString cell_name(cell);
+                log_info("      - %s\n", cell_name.c_str(ctx));
+            }
+        log_info("  - patterns:\n");
+            for (auto pattern : bel_chain.chain_patterns) {
+                IdString source(pattern.source);
+                IdString sink(pattern.sink);
+                log_info("      - %s -> %s\n", source.c_str(ctx), sink.c_str(ctx));
+            }
+        log_info("  - coord_configs:\n");
+            for (auto cfg : bel_chain.chain_coord_configs) {
+                log_info("      - coord: %d | step: %d\n", cfg.coord, cfg.step);
+            }
+    }
+}
+
+void Arch::prepare_cluster(const BelChainPOD *chain)
+{
+    Context *ctx = getCtx();
+
+    // Get chainable cells
+    std::vector<CellInfo *> chainable_cells;
+    for (auto cell : sorted(ctx->cells)) {
+        CellInfo *ci = cell.second;
+        for (auto chain_cell : chain->cells) {
+            IdString chain_cell_type(chain_cell);
+            if (ci->type == chain_cell_type) {
+                chainable_cells.push_back(ci);
+            }
+        }
+    }
+
+    // Prepare sources and sinks pairs from provided patterns
+    // pair: <cell_type, cell_port>
+    std::vector<std::pair<IdString, IdString>> sources, sinks;
+    std::string delimiter = ".";
+    for (auto pattern : chain->chain_patterns) {
+        IdString source(pattern.source);
+        IdString sink(pattern.sink);
+        IdString source_type;
+        IdString sink_type;
+        std::string source_str = source.c_str(ctx);
+        std::string sink_str = sink.c_str(ctx);
+        std::string source_type_str;
+        std::string sink_type_str;
+
+        source_type_str = source_str.substr(0, source_str.find(delimiter));
```

Comment:
it might be a bit cleaner to do the splitting, and conversion to IdString, inside the bba generation rather than at runtime